Modules/_asynciomodule.c

```diff
@@ -3564,55 +3751,32 @@ static PyObject *
 _asyncio_all_tasks_impl(PyObject *module, PyObject *loop)
 /*[clinic end generated code: output=0e107cbb7f72aa7b input=43a1b423c2d95bfa]*/
 {
-
     asyncio_state *state = get_asyncio_state(module);
-    PyObject *tasks = PySet_New(NULL);
-    if (tasks == NULL) {
-        return NULL;
-    }
     if (loop == Py_None) {
         loop = _asyncio_get_running_loop_impl(module);
         if (loop == NULL) {
-            Py_DECREF(tasks);
             return NULL;
         }
     } else {
         Py_INCREF(loop);
     }
     // First add eager tasks to the set so that we don't miss
     // any tasks which graduates from eager to non-eager
-    PyObject *eager_iter = PyObject_GetIter(state->eager_tasks);
-    if (eager_iter == NULL) {
-        Py_DECREF(tasks);
-        Py_DECREF(loop);
-        return NULL;
-    }
-    PyObject *item;
-    while ((item = PyIter_Next(eager_iter)) != NULL) {
-        if (add_one_task(state, tasks, item, loop) < 0) {
-            Py_DECREF(tasks);
-            Py_DECREF(loop);
-            Py_DECREF(item);
-            Py_DECREF(eager_iter);
-            return NULL;
-        }
-        Py_DECREF(item);
-    }
-    Py_DECREF(eager_iter);
-
-    if (PyErr_Occurred()) {
-        Py_DECREF(tasks);
+    // We first add all the tasks to `tasks` set and then filter
+    // out the tasks which are done and return it.
+    PyObject *tasks = PySet_New(state->eager_tasks);
```

Comment:
I think the strategy of collecting everything in `tasks` before processing it makes sense, but I think we should use a list here instead of a set. That will avoid a number of potential reentrancy issues due to `PySet_Add()` calling potentially escaping hash and equality functions.

If you need to de-duplicate elements, either do it as part of `add_one_task` or convert the list to a set in a single step once you've collected all the tasks.

Reply:
Yeah that sounds better, I have changed it to now use a list